pretyping/canonicalStructures.mli

```diff
@@ -0,0 +1,156 @@
+val register_primitive_projection : Names.Projection.Repr.t -> Names.Constant.t -> unit
+
+val is_primitive_projection : Names.Constant.t -> bool
+
+val find_primitive_projection : Names.Constant.t -> Names.Projection.Repr.t option
+
+(** {6 Canonical structures } *********************************************** *)
+
+(** A canonical structure declares "canonical" conversion hints between
+    the effective components of a structure and the projections of the
+    structure *)
+
+(** A ValuePattern.t characterizes the form of a component of canonical structure *)
+
+module ValuePattern : sig
+type t =
+    Const_cs of Names.GlobRef.t
+  | Proj_cs of Names.Projection.Repr.t
+  | Prod_cs
+  | Sort_cs of Sorts.family
+  | Default_cs
+
+val equal : Environ.env -> t -> t -> bool
+val print : t -> Pp.t
+
+(** Return the form of the component of a canonical structure *)
+val of_constr : Environ.env -> constr -> t * int option * constr list
+
+end
+
+type cs
+
+val check_and_decompose_canonical_structure :
+  Environ.env -> Evd.evar_map -> Names.GlobRef.t -> cs
+
+val register_canonical_structure :
+  warn:bool -> Environ.env -> Evd.evar_map -> cs -> unit
+
+val subst_canonical_structure : Mod_subst.substitution -> cs -> cs
+val solution_of_canonical_structure : cs -> Names.GlobRef.t
+
+(** The canonical solution of a problem (proj,val) is a global
+    [constant = fun abs : abstractions_ty => body] and
+    [body = RecodConstructor params canon_values] and the canonical value
+    corresponding to val is [val cvalue_arguments].
+    It is possible that val is one of the [abs] abstractions, eg [Default_cs],
+    and in that case [cvalue_abstraction = Some i] *)
+
+module Solution : sig
+type t = {
+  constant : EConstr.t;
+
+  abstractions_ty : EConstr.t list;
+  body : EConstr.t;
+
+  nparams : int;
+  params : EConstr.t list;
+  cvalue_abstraction : int option;
+  cvalue_arguments : EConstr.t list;
+}
+
+end
+
+(** [lookup_canonical_conversion (p,v)] returns a s such that p s = v.
+    The solution s gets a fresh universe instance and is decomposed into
+    bits for consumption by evarconv *)
+val lookup_canonical_conversion :
+  Environ.env -> Evd.evar_map -> (Names.GlobRef.t * ValuePattern.t) ->
+    Evd.evar_map * Solution.t
+
+(** [is_open_canonical_projection env sigma t] is true if t is a FieldName
+    applied to term which is not a constructor *)
+val is_open_canonical_projection :
+  Environ.env -> Evd.evar_map -> EConstr.t -> bool
+
+(** Access to the Canonical Structure database *)
+module TableEntry : sig
+
+type t = {
+  projection : Names.GlobRef.t;
+  value : ValuePattern.t;
+  solution : Names.GlobRef.t;
+}
+
+end
+
+(** [canonical_entries] returns the list of tuples { p ; v ; s }
+   Note: p s = v *)
+val canonical_entries : unit -> TableEntry.t list
+
+(** [canonical_entries_for p] returns the list of canonical entries that have
+    p as their FieldName *)
+val canonical_entries_for : projection:Names.GlobRef.t -> TableEntry.t list
```

Comment:
Same comment, maybe you wanna place some functions inside the module they pertain.